File: backend/app/schemas/workflow.py
"""
工作流 App / 审批 — Pydantic schemas
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ── 设计态 CRUD ──────────────────────────────────────────

class AppCreate(BaseModel):
    code: str = Field(..., max_length=100)
    name: str = Field(..., max_length=200)
    description: str | None = None
    scene_code: str | None = None
    canvas_json: dict[str, Any] | None = None


class AppUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    scene_code: str | None = None
    canvas_json: dict[str, Any] | None = None


class AppBrief(BaseModel):
    id: str
    code: str
    name: str
    description: str | None
    status: str
    scene_code: str | None
    published_version: int
    created_at: datetime | None
    updated_at: datetime | None

    class Config:
        from_attributes = True


class AppDetail(AppBrief):
    canvas_json: dict[str, Any] | None
    published_json: dict[str, Any] | None

    class Config:
        from_attributes = True


# ── 发布 ─────────────────────────────────────────────────

class PublishResult(BaseModel):
    app_id: str
    version: int
    published_json: dict[str, Any]


# ── 工具注册 ─────────────────────────────────────────────

class ToolSpec(BaseModel):
    name: str
    description: str
    sensitive: bool = False
    parameters: dict[str, Any] = {}
    required: list[str] = []


# ── 审批 ─────────────────────────────────────────────────

class ApprovalBrief(BaseModel):
    id: str
    app_id: str
    session_id: str
    tool_name: str
    tool_args: dict[str, Any] | None
    status: str
    created_by: str
    resolved_by: str | None
    created_at: datetime | None
    resolved_at: datetime | None

    class Config:
        from_attributes = True


class ApprovalAction(BaseModel):
    comment: str | None = None

File: frontend/src/types/workflow.ts
// ── 设计态 (Design-Time) ─────────────────────────────────

export interface DesignNodeData {
  nodeType: string
  label: string
  color: string
  icon: string
  description?: string
  config: Record<string, any>
}

export interface WorkflowApp {
  id: string
  code: string
  name: string
  description: string | null
  status: 'draft' | 'published'
  scene_code: string | null
  published_version: number
  canvas_json: { nodes: any[]; edges: any[]; viewport?: any } | null
  published_json: PublishedApp | null
  created_at: string | null
  updated_at: string | null
}

export interface AppBrief {
  id: string
  code: string
  name: string
  description: string | null
  status: string
  scene_code: string | null
  published_version: number
  created_at: string | null
  updated_at: string | null
}

export interface AppCreate {
  code: string
  name: string
  description?: string
  scene_code?: string
  canvas_json?: Record<string, any>
}

export interface AppUpdate {
  name?: string
  description?: string
  scene_code?: string
  canvas_json?: Record<string, any>
}

// ── 运行态 (Run-Time) ───────────────────────────────────

export interface PublishedApp {
  appId: string
  appCode: string
  appName: string
  sceneCode: string | null
  ontologyScope: string[]
  defaultAgent: AgentConfig | null
  toolBindings: ToolBinding[]
  approvalPolicy: ApprovalPolicy
  widgetBindings: WidgetBinding[]
  version: number
}

export interface AgentConfig {
  nodeId: string
  persona: string
  objective: string
  maxSteps: number
  boundTools: string[]
}

export interface ToolBinding {
  name: string
  description: string
  sensitive: boolean
  parameters: Record<string, any>
  required: string[]
}

export interface ApprovalPolicy {
  sensitiveTools: string[]
  requireApproval: boolean
}

export interface WidgetBinding {
  widgetType: string
  nodeId: string
  label: string
  boundAgentNodeId: string | null
}

// ── 工具注册 ─────────────────────────────────────────────

export interface ToolSpec {
  name: string
  description: string
  sensitive: boolean
  parameters: Record<string, any>
  required: string[]
}

// ── 审批 ─────────────────────────────────────────────────

export interface ApprovalTask {
  id: string
  app_id: string
  session_id: string
  tool_name: string
  tool_args: Record<string, any> | null
  status: 'pending' | 'approved' | 'rejected'
  created_by: string
  resolved_by: string | null
  created_at: string | null
  resolved_at: string | null
}

// ── 运行时上下文 ─────────────────────────────────────────

export interface ChatRuntimeContext {
  sceneCode: string
  ontologyType: string
  objectId?: string
  tenantId?: string
  userId?: string
  extra?: Record<string, any>
}
